### PR TITLE
Implement full YARPC interface.

### PR DESCRIPTION
**Pull request by abhinav**
Depends on #11.

examples/json/server.go can now run and process requests.

```
$ curl -X POST -H 'YARPC-Procedure: set' http://localhost:8080 -d '{"key": "hello", "value": "world"}'
{}
$ curl -X POST -H 'YARPC-Procedure: get' http://localhost:8080 -d '{"key": "hello"}'
{"value":"world"}
```

### CODE REVIEW

**Comment by abhinav** [comment]
@breerly @akshayjshah @prashantv @kriskowal

File: errors.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+type errorGroup struct {
+	Errors []error
+}
+
+func (e errorGroup) Error() string {
+	messages := []string{"the following errors occurred:"}
+	for _, err := range e.Errors {
+		messages = append(messages, "\t"+err.Error())
+	}
+	return strings.Join(messages, "\n")
```
Comment (prashantv) [inline errors.go]: don't need to prepend "\t" to the error messages if you have a join here, can probably just do
```go
return strings.Join(messages, "\n\t")
```

File: examples/json/server.go [hunk 2]
```diff
@@ -68,4 +62,9 @@ func main() {
 	handler := handler{items: make(map[string]string)}
 	json.Register(yarpc, json.Procedure("get", handler.Get))
 	json.Register(yarpc, json.Procedure("set", handler.Set))
+
+	err := yarpc.Serve()
+	if err != nil {
```
Comment (prashantv) [inline examples/json/server.go]: nit: inline `Serve` call in if

File: rpc.go [hunk 3]
```diff
@@ -20,25 +20,116 @@
...
+			return i.Serve(r)
+		}
+	}
+
+	var wait sync.ErrorWaiter
+	for _, i := range r.Inbounds {
+		wait.Submit(callServe(i))
+	}
+
+	errors := wait.Wait()
```
Comment (prashantv) [inline rpc.go]: nit: can inline into if below

File: rpc.go [hunk 4]
```diff
@@ -20,25 +20,116 @@
...
+	return h.Handle(ctx, req)
+}
+
+func (r rpc) Close() error {
+	var wait sync.ErrorWaiter
+	for _, i := range r.Inbounds {
+		wait.Submit(i.Close)
+	}
+
+	errors := wait.Wait()
```
Comment (prashantv) [inline rpc.go]: nit: can inline into if below

File: sync/err.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+// Submit submits a task for executino on the ErrorWaiter.
+//
+// The function returns immediately.
+func (ew *ErrorWaiter) Submit(f func() error) {
+	ew.wait.Add(1)
+	go func() {
+		defer ew.wait.Done()
+		err := f()
```
Comment (prashantv) [inline sync/err.go]: nit: inline into if below

File: errors.go (hunk 1 above)
Comment (abhinav) [inline errors.go]: ah true

File: examples/json/server.go (hunk 2 above)
Comment (abhinav) [inline examples/json/server.go]: :+1:

File: rpc.go (hunk 3 above)
Comment (abhinav) [inline rpc.go]: :+1:

File: rpc.go (hunk 4 above)
Comment (abhinav) [inline rpc.go]: :+1:

File: sync/err.go (hunk 5 above)
Comment (abhinav) [inline sync/err.go]: :+1:

**Comment by abhinav** [comment]
Updatde

**Comment by prashantv** [comment]
lgtm. Good idea with the ErrorWaiter pattern, looks pretty nice.

File: errors.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+
+import "strings"
+
+// errorGroup represents a collection of errors.
+type errorGroup struct {
+	Errors []error
+}
+
+func (e errorGroup) Error() string {
+	messages := []string{"the following errors occurred:"}
```
Comment (akshayjshah) [inline errors.go]: nit: would this be a bit better as

```
messages := make([]string, 0, len(e.Errors))
messages = append(messages, "the following errors occurred:")
```

File: rpc.go [hunk 7]
```diff
@@ -20,25 +20,114 @@
...
 type RPC interface {
+	transport.Handler
 	transport.Registry
 
 	// Retrieves a new Outbound transport that will make requests to the given
 	// service.
+	//
+	// This panics if the given service is unknown.
 	Channel(service string) transport.Outbound
+	// TODO do we really want to panic on unknown services?
```
Comment (akshayjshah) [inline rpc.go]: :no_good: Rather not. At some point in the future, perhaps we can default to a stub that returns the YARPC version of 404s for all procedure calls.

File: rpc.go [hunk 8]
```diff
@@ -20,25 +20,114 @@
...
+}
+
+func (r rpc) Serve() error {
+	callServe := func(i transport.Inbound) func() error {
+		return func() error {
+			return i.Serve(r)
+		}
+	}
+
+	var wait sync.ErrorWaiter
```
Comment (akshayjshah) [inline rpc.go]: Immediately makes me think of 
image

This is a pretty slick interface, though - nice!

Comment (breerly) [inline rpc.go]: Go with the original OG:

butler1000

File: scheme/json/errors.go [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package json
+
+import "fmt"
+
+// unmarshalError is raised when there's an error parsing JSON input.
```
Comment (akshayjshah) [inline scheme/json/errors.go]: Should we use the term "raised" in Go? It's not an exception, so the term feels a bit out of place.

File: sync/err.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+
+// ErrorWaiter is similar to a WaitGroup except it allows collecting failures
+// from subtasks.
+type ErrorWaiter struct {
+	wait   sync.WaitGroup
+	lock   sync.Mutex
+	errors []error
+}
+
+// Submit submits a task for executino on the ErrorWaiter.
```
Comment (akshayjshah) [inline sync/err.go]: spelling - "execution"

File: errors.go (hunk 6 above)
Comment (abhinav) [inline errors.go]: Updated

File: scheme/json/errors.go (hunk 9 above)
Comment (abhinav) [inline scheme/json/errors.go]: fair. changed

File: sync/err.go (hunk 10 above)
Comment (abhinav) [inline sync/err.go]: :+1:

**Pull request merged**